src/Generic/Data/Surgery/Internal.hs

```diff
@@ -866,6 +953,80 @@ type Arborify' arb op n nDiv2 f g =
    <=< SplitAt nDiv2
    ) (op f g)
 
+class GLazify f where
+  gLazify :: f x -> Lazify f x
+
+instance GLazify f => GLazify (M1 i m f) where
+  gLazify (M1 a) = M1 (gLazify a)
+
+instance (GLazify f, GLazify g) => GLazify (f :*: g) where
+  gLazify (a :*: b) = gLazify a :*: gLazify b
+
+instance (GLazify f, GLazify g) => GLazify (f :+: g) where
+  gLazify = \case
+    L1 a -> L1 (gLazify a)
+    R1 b -> R1 (gLazify b)
+
+instance GLazify (K1 i c) where
+  gLazify = id
+
+instance GLazify U1 where
+  gLazify = id
+
+instance GLazify V1 where
+  gLazify = id
+
+type family Lazify (f :: k -> *) :: k -> *
+type instance Lazify (M1 i m f) = M1 i (LazifyMeta m) (Lazify f)
+type instance Lazify (f :*: g) = Lazify f :*: Lazify g
+type instance Lazify (f :+: g) = Lazify f :+: Lazify g
+type instance Lazify (K1 i c) = K1 i c
+type instance Lazify U1 = U1
+type instance Lazify V1 = V1
+
+type family LazifyMeta (m :: Meta) :: Meta
+type instance LazifyMeta ('MetaData n m p nt) = 'MetaData n m p nt
+type instance LazifyMeta ('MetaCons n f s) = 'MetaCons n f s
+type instance LazifyMeta ('MetaSel mn su ss ds)
+  = 'MetaSel mn 'NoSourceUnpackedness 'NoSourceStrictness 'DecidedLazy
+
+class GUnLazify from to where
+  gUnLazify :: to x -> UnLazify from to x
+
+instance GUnLazify f f' => GUnLazify (M1 i m f) (M1 i m' f') where
+  gUnLazify (M1 a) = M1 (gUnLazify @f @f' a)
+
+instance (GUnLazify f f', GUnLazify g g') => GUnLazify (f :*: g) (f' :*: g') where
+  gUnLazify (a :*: b) = gUnLazify @f @f' a :*: gUnLazify @g @g' b
+
+instance (GUnLazify f f', GUnLazify g g') => GUnLazify (f :+: g) (f' :+: g') where
+  gUnLazify = \case
+    L1 a -> L1 (gUnLazify @f @f' a)
+    R1 b -> R1 (gUnLazify @g @g' b)
+
+instance GUnLazify (K1 i c) (K1 i c) where
+  gUnLazify = id
+
+instance GUnLazify U1 U1 where
+  gUnLazify = id
+
+instance GUnLazify V1 V1 where
+  gUnLazify = id
+
+type family UnLazify (from :: k -> *) (to :: k -> *) :: k -> *
+type instance UnLazify (M1 i m f) (M1 i m' f') = M1 i (UnLazifyMeta m m') (UnLazify f f')
+type instance UnLazify (f :*: g) (f' :*: g') = UnLazify f f' :*: UnLazify g g'
+type instance UnLazify (f :+: g) (f' :+: g') = UnLazify f f' :+: UnLazify g g'
+type instance UnLazify (K1 i c) (K1 i c)  = K1 i c
+type instance UnLazify U1 U1 = U1
+type instance UnLazify V1 V1 = V1
```

Comment:
The `Lazify` type family can also be used in the other direction instead of an additional `UnLazify`.

Reply:
Not sure how to replace `UnLazify` with `Lazify`, given that we want to account for any possible shape of `MetaSel` - could you give me a nudge in the right direction?

Reply:
My idea is that instead of `f ~ UnLazify f (Arborify l)` we can write `Lazify f ~ Arborify l`.

Note also that in `fromOR` (and `fromLazyOR`) we can really only infer `l` from `a`, not the other way around. So the most useful constraint would be `Linearize (Lazify f) ~ l`, since that's what allows that inference to happen (assuming `f ~ Rep a`).

We can still keep `Lazify f ~ Arborify l` around because that would allow `fromLazyOR` to work in the case where `a` is a `Data f x`. I'm not sure where that's actually useful (if you know you want a `Data`, the more precise type of `fromOR'` helps type inference even further), but it seems nice to have without costing much.

Reply:
I had completely missed this, thank you very much for all the interesting insights!